include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -113,30 +139,74 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
 
-        const _Size __n_items = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __iters_per_work_item);
+        _Size __size_per_work_group =
+            __iters_per_work_item1 * __work_group_size; // number of buffer elements processed within workgroup
+        _Size __n_groups = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __size_per_work_group);
+        _Size __n_items = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __iters_per_work_item1);
 
-        sycl::buffer<_Tp> __res(sycl::range<1>(1));
+        sycl::buffer<_Tp> __temp(sycl::range<1>(1 * __n_groups));
 
         sycl::event __reduce_event = __exec.queue().submit([&, __n, __n_items](sycl::handler& __cgh) {
             oneapi::dpl::__ranges::__require_access(__cgh, __rngs...); // get an access to data under SYCL buffer
-            auto __res_acc = __res.template get_access<access_mode::write>(__cgh);
+            auto __temp_acc = __temp.template get_access<sycl::access_mode::write>(__cgh);
             __dpl_sycl::__local_accessor<_Tp> __temp_local(sycl::range<1>(__work_group_size), __cgh);
-            __cgh.parallel_for<_Name...>(
-                sycl::nd_range<1>(sycl::range<1>(__work_group_size), sycl::range<1>(__work_group_size)),
+            __cgh.parallel_for<_MainName...>(
+                sycl::nd_range<1>(sycl::range<1>(__n_groups * __work_group_size), sycl::range<1>(__work_group_size)),
                 [=](sycl::nd_item<1> __item_id) {
-                    ::std::size_t __global_idx = __item_id.get_global_id(0);
-                    ::std::uint16_t __local_idx = __item_id.get_local_id(0);
+                    auto __local_idx = __item_id.get_local_id(0);
+                    auto __group_idx = __item_id.get_group(0);
                     // 1. Initialization (transform part). Fill local memory
-                    __transform_pattern(__local_idx, __n, __iters_per_work_item, __global_idx, /*global_offset*/ 0,
-                                        __temp_local, __rngs...);
+                    __transform_pattern1(__item_id, __n, /*global_offset*/ 0, __temp_local, __rngs...);
                     __dpl_sycl::__group_barrier(__item_id);
                     // 2. Reduce within work group using local memory
-                    _Tp __result = __reduce_pattern(__item_id, __global_idx, __n_items, __temp_local);
+                    _Tp __result = __reduce_pattern(__item_id, __n_items, __temp_local);
+                    if (__local_idx == 0)
+                        __temp_acc[__group_idx] = __result;
+                });
+        });
+
+        __n = __n_groups;
+
+        // Lower the work group size of the second kernel to the next power of 2 if __n < __work_group_size.
+        auto __work_group_size2 = __work_group_size;
+        if constexpr (__iters_per_work_item2 == 1)
+        {
+            if (__n < __work_group_size)
+            {
+                __work_group_size2 = __n;
+                if ((__work_group_size2 & (__work_group_size2 - 1)) != 0)
```

Comment:
I see in our codebase that we have `__dpl_bit_floor` but no equivalent `__dpl_bit_ceil`. Am I correct that this if statement, bit floor, and bit shift could be replaced with a single bit ceil operation?

Since `std::bit_ceil` is in the C++20 standard (https://en.cppreference.com/w/cpp/numeric/bit_ceil), it may be useful to add an equivalent `__dpl_bit_ceil` into our utilities.

Reply:
This is correct. I will change that to `__dpl_bit_ceil ` once #909 is merged.